engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala

```diff
@@ -6357,6 +6360,45 @@ object IR {
 
     /** The location at which the diagnostic occurs. */
     val location: Option[IdentifiedLocation]
+
+    /** Equals based on returned keys.
+      */
+    override def equals(other: Any): Boolean = {
```

Comment:
I'm very uneasy about changing `equals` like this. Particularly with scala's `case class` it is understood that equality and hashCode do what they always do (which is structural equality). And this offers a weird notion of equality instead.

Reply:
A clean solution here would be to move this notion of equality to a separate wrapper object that _only_ exposes this notion and is very obvious about it (i.e. there's no way someone accidentally uses it)

Reply:
OK, let's encapsulate the `equals` & `hashCode` change a bit: [af0c659](https://github.com/enso-org/enso/pull/3465/commits/af0c659a22442ebefbf82a0330fbaae55a2872bb)